File: objc/Project4/Project4 WatchKit Extension/ResultsInterfaceController.m
//
//  ResultsInterfaceController.m
//  Project4 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/24/21.
//

#import "ResultsInterfaceController.h"

@implementation ResultsInterfaceController

- (void)awakeWithContext:(id)context {
    [super awakeWithContext:context];
    
    NSDictionary<NSString *, NSString *> *settings = context;
    if (settings == nil) return;
    NSString *amount = settings[@"amount"];
    if (amount == nil) return;
    NSString *baseCurrency = settings[@"base"];
    if (baseCurrency == nil) return;
    
    self.amountToConvert = [amount floatValue];
    [self setTitle:[NSString stringWithFormat:@"%@ %@", amount, baseCurrency]];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self fetchDataFor:baseCurrency];
    });
}

- (void)willActivate {
    [super willActivate];
    [WKExtension.sharedExtension setAutorotating:YES];
}

- (void)didDeactivate {
    [super didDeactivate];
    [WKExtension.sharedExtension setAutorotating:NO];
}

- (IBAction)doneTapped {
    [WKInterfaceController reloadRootPageControllersWithNames:@[@"Home", @"Currencies"]
                                                     contexts:nil
                                                  orientation:WKPageOrientationHorizontal
                                                    pageIndex:0];
}

- (void)fetchDataFor:(NSString *)baseCurrency {
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://openexchangerates.org/api/latest.json?app_id=%@&base=%@", appID, baseCurrency]];
    if (url) {
        NSURLRequest *urlRequest = [[NSURLRequest alloc] initWithURL:url];
        
        [[NSURLSession.sharedSession dataTaskWithRequest:urlRequest completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error){
            if (data) {
                [self parse:data];
            } else {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.status setText:@"Fetch failed"];
                    [self.done setHidden:NO];
                });
            }
        }] resume];
    } else {
        [self.status setText:@"Fetch failed"];
        [self.done setHidden:NO];
    }
}

- (void)parse:(NSData *)contents {
    NSError * _Nullable jsonError = nil;
    NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:contents options:0 error:&jsonError];
    if ((jsonError != nil) || (dic == nil)) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.status setText:@"Fetch failed"];
            [self.done setHidden:NO];
        });
        return;
    }
    
    // load their currency selection
    NSUserDefaults *defaults = NSUserDefaults.standardUserDefaults;
    NSArray *selectedCurrencies = [defaults arrayForKey:[InterfaceController selectedCurrenciesKey]];
    if (selectedCurrencies == nil) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.status setText:@"Fetch failed"];
            [self.done setHidden:NO];
        });
        return;
    }
    
    NSDictionary<NSString *, NSNumber *> *rates = dic[@"rates"];
    if (rates == nil) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.status setText:@"Fetch failed"];
            [self.done setHidden:NO];
        });
        return;
    }
    
    NSMutableDictionary<NSString *, NSNumber *> *fetched = [@{} mutableCopy];
    [rates enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull currency, NSNumber * _Nonnull rate, BOOL * _Nonnull stop){
        // only include currencies the user wants
        if (![selectedCurrencies containsString:currency]) return;
        [fetched addEntriesFromDictionary:@{currency: rate}];
    }];
    self.fetchedCurrencies = fetched;
    
    [self updateTable];
    [self.status setHidden:YES];
    [self.table setHidden:NO];
    [self.done setHidden:NO];
}

- (void)updateTable {
    // load as many rows as we have currencies
    [self.table setNumberOfRows:self.fetchedCurrencies.count withRowType:@"Row"];
    
    // loop over each cuurency, getting its position and symbol
    NSUInteger __block index = 0;
    [self.fetchedCurrencies enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull currency, NSNumber * _Nonnull rate, BOOL * _Nonnull stop){
        // find the row at that position
        CurrencyRow *row = [self.table rowControllerAtIndex:index];
        if (row == nil) *stop = YES;
        
        // multiply the rate by the user's input amount
        float value = self.amountToConvert * [rate floatValue];
        
        // convert it to a rounded string
        NSString *formattedValue = [NSString stringWithFormat:@"%.2f", value];
        
        // show it in the label
        [row.textLabel setText:[NSString stringWithFormat:@"%@ %@", formattedValue, currency]];
        index += 1;
    }];
}

@end

File: swift/Project4/Project4 WatchKit Extension/ResultsInterfaceController.swift
//
//  ResultsInterfaceController.swift
//  Project4 WatchKit Extension
//
//  Created by Jinwoo Kim on 1/24/21.
//

import WatchKit
import Foundation


class ResultsInterfaceController: WKInterfaceController {
    @IBOutlet weak var table: WKInterfaceTable!
    @IBOutlet weak var status: WKInterfaceLabel!
    @IBOutlet weak var done: WKInterfaceButton!
    
    var fetchedCurrencies = [(symbol: String, rate: Double)]()
    var amountToConvert = 0.0
    let appID = "051bce45cd804955bfdfedbab80dd347"
    
    override func awake(withContext context: Any?) {
        super.awake(withContext: context)
        
        guard let settings = context as? [String: String] else { return }
        guard let amount = settings["amount"] else { return }
        guard let baseCurrency = settings["base"] else { return }
        
        amountToConvert = Double(amount) ?? 50
        setTitle("\(amount) \(baseCurrency)")
        
        DispatchQueue.global(qos: .userInteractive).async {
            self.fetchData(for: baseCurrency)
        }
    }
    
    override func willActivate() {
        super.willActivate()
        WKExtension.shared().isAutorotating = true
    }
    
    override func didDeactivate() {
        super.didDeactivate()
        WKExtension.shared().isAutorotating = false
    }
    
    @IBAction func doneTapped() {
        WKInterfaceController.reloadRootPageControllers(withNames: ["Home", "Currencies"], contexts: nil, orientation: .horizontal, pageIndex: 0)
    }
    
    func fetchData(for baseCurrency: String) {
        if let url = URL(string: "https://openexchangerates.org/api/latest.json?app_id=\(appID)&base=\(baseCurrency)") {
            let urlRequest = URLRequest(url: url)
            
            URLSession.shared.dataTask(with: urlRequest) { data, response, error in
                if let data = data {
                    self.parse(data)
                } else {
                    DispatchQueue.main.async {
                        self.status.setText("Fetch failed")
                        self.done.setHidden(false)
                    }
                }
            }.resume()
        }
    }
    
    func parse(_ contents: Data) {
        let decoder = JSONDecoder()
        
        guard let result = try? decoder.decode(CurrencyResult.self, from: contents) else {
            // we failed to decode - show an error!
            DispatchQueue.main.async {
                self.status.setText("Fetch failed")
                self.done.setHidden(false)
            }
            
            return
        }
        
        // load their currency selection
        let defaults = UserDefaults.standard
        let selectedCurrencies = defaults.array(forKey: InterfaceController.selectedCurrenciesKey) as? [String] ?? InterfaceController.defaultCurrencies
        
        for symbol in result.rates {
            // only include currencies the user wants
            guard selectedCurrencies.contains(symbol.key) else { continue }
            let rateName = symbol.key
            let rateValue = symbol.value
            fetchedCurrencies.append((symbol: rateName, rate: rateValue))
        }
        
        updateTable()
        status.setHidden(true)
        table.setHidden(false)
        done.setHidden(false)
    }
    
    func updateTable() {
        // load as many rows as we have currencies
        table.setNumberOfRows(fetchedCurrencies.count, withRowType: "Row")
        
        // loop over each currency, getting its position and symbol
        for (index, currency) in fetchedCurrencies.enumerated() {
            // find the row at that position
            guard let row = table.rowController(at: index) as? CurrencyRow else { return }
            
            // multiply the rate by the user's input amount
            let value = currency.rate * amountToConvert
            
            // convert it to a rounded string
            let formattedValue = String(format: "%.2f", value)
            
            // show it in the label
            row.textLabel.setText("\(formattedValue) \(currency.symbol)")
        }
    }
}
